### PR TITLE
Add more context when failing to unmarshal JSON

### LINKED ISSUE #830: Occasional invalid JSON returned by Stripe
Very occasionally (once every few hundred thousands requests) I will receive an apparently nonsensical error from `charge.New()`. Errors such as:

```
invalid character 'o' in literal null (expecting 'u')
```
and
```
invalid character 'u' looking for beginning of value
```

After a bit of head-scratching I realised that these are actually JSON unmarshaling errors so I had a look in `stripe.go` and found this at line 437:

```go
	if res.StatusCode >= 400 {
		return s.ResponseToError(res, resBody)
	}

	s.LeveledLogger.Debugf("Response: %s\n", string(resBody))

	if v != nil {
		return json.Unmarshal(resBody, v)
	}

	return nil
```

and in ResponseToError at line 451:

```go
func (s *BackendImplementation) ResponseToError(res *http.Response, resBody []byte) error {
	var raw rawError
	if err := json.Unmarshal(resBody, &raw); err != nil {
		return err
	}
```

The problem here is that invalid JSON is being returned from the remote API but the client code is making no effort at exposing it. And if the response is a 400 or worse, it doesn't even log the actual body to the debug log.

It would be useful to be able to get the actual raw response from the request so it can be logged and we can understand why we're seeing failures, rather than just cryptic notes about unexpected vowels.

### PR DESCRIPTION
**Pull request by brandur**
As mentioned in #830, we occasionally have a situation where non-JSON
data comes back to a client even though this is not supposed to happen.
Unfortunately, when it does, the error message produced by Go isn't very
good and doesn't help much with debugging.

Here we add a new helper function for unmarshaling JSON in standard or
error responses that behaves normally in the normal cases, and in case
of a failure, adds more context to the standard library's error and
logs.

Fixes #830.

r? @remi-stripe

### CODE REVIEW

File: stripe.go [hunk 1]
```diff
@@ -505,6 +506,32 @@ func (s *BackendImplementation) SetNetworkRetriesSleep(sleep bool) {
...
+	err := json.Unmarshal(body, v)
+	if err != nil {
+		// If we got invalid JSON back then something totally unexpected is
+		// happening (caused by a bug on the server side). Put a sample of the
+		// response body into the error message so we can get a better feel for
+		// what the problem was.
+		bodySample := string(body)
+		if len(bodySample) > 50 {
+			bodySample = bodySample[0:50]
+		}
```
[the remark this answers is not in the document]
Comment (brandur-stripe) [reply stripe.go]: Yeah it's possible, but I'm basically trying to protect against a totally degenerate scenario where we say dump a 40k HTML page or something by accident.

I think we should pick a number and truncate at it as protection, but I'm okay with one that's on the bigger side though, like 500 or such. What do you think?

**Comment by brandur-stripe** [comment]
ptal @remi-stripe

**Comment by brandur-stripe** [comment]
I just upped the sample size from 50 to 500.

**Comment by remi-stripe** [comment]
@brandur-stripe Is there a way to access the raw body/response in the library like we do in other libs?

**Comment by brandur-stripe** [comment]
> @brandur-stripe Is there a way to access the raw body/response in the library like we do in other libs?

@remi-stripe You mean from the user's perspective right?
No, I don't think so.

We could consider putting something like that in though.
Creating a "response metadata" struct and embedding it in
objects that get returned from API methods might work:

``` go
type StripeResponseData struct {
	Body      []byte
	RequestID string
}

// Charge is the resource representing a Stripe charge.
// For more details see https://stripe.com/docs/api#charges.
type Charge struct {
	StripeResponseData

	Amount               int64                       `json:"amount"`
	AmountRefunded       int64                       `json:"amount_refunded"`
	Application          *Application                `json:"application"`
	ApplicationFee       *ApplicationFee             `json:"application_fee"`
    ...
}
```

**Comment by remi-stripe** [comment]
Ack. Yeah that's what I meant because when we debug those we almost always ask for the raw response received. But we can always cross that bridge if we need this in the future!

**Comment by brandur-stripe** [comment]
> But we can always cross that bridge if we need this in the future!

SG. Thanks Remi!

**Pull request merged**